drivers/net/dsa/lantiq_gsw_mdio.c

```diff
@@ -35,19 +36,103 @@
 #include "lantiq_gsw.h"
 #include "lantiq_pce.h"
 
-static u32 gsw_mdio_read(struct gswip_priv *priv, void *address)
+#define NUM_ACCESSIBLE_REGS (30)
+#define TARGET_BASE_ADDRESS_REG (31)
+
+struct gsw_mdio {
+	struct mdio_device *mdio_dev;
+	struct gswip_priv common;
+};
+
+static u32 gsw_mdio_read_actual(struct mdio_device *mdio, u32 reg)
 {
-	/* TODO WARP-5829: write this function */
-	return 0;
+	return mdio->bus->read(mdio->bus, mdio->addr, reg);
 }
 
-static void gsw_mdio_write(struct gswip_priv *priv, u32 val, void *address)
+static void gsw_mdio_write_actual(struct mdio_device *mdio, u32 reg, u32 val)
 {
-	/* TODO WARP-5829: write this function */
+	mdio->bus->write(mdio->bus, mdio->addr, reg, val);
+}
+
+static inline u32 gsw_mdio_read_tbar(struct mdio_device *mdio)
+{
+	return gsw_mdio_read_actual(mdio, TARGET_BASE_ADDRESS_REG);
+}
+
+static inline void gsw_mdio_write_tbar(struct mdio_device *mdio, u32 reg_addr)
+{
+	gsw_mdio_write_actual(mdio, TARGET_BASE_ADDRESS_REG, reg_addr);
+}
+
+static u32 gsw_mdio_check_write_tbar(struct mdio_device *mdio, u32 reg_addr)
+{
+	u32 tbar = gsw_mdio_read_tbar(mdio);
+
+	/* MDIO slave interface uses an indirect addressing scheme that allows
+	 * access to NUM_ACCESSIBLE_REGS registers at a time. The Target Base
+	 * Address Register (TBAR) is used to set a base offset, then MDIO
+	 * registers (0-30) are used to access internal addresses of (TBAR + 0-30)
+	 */
+	if ((reg_addr < tbar) || (reg_addr > (tbar + NUM_ACCESSIBLE_REGS))) {
+			gsw_mdio_write_tbar(mdio, reg_addr);
+			tbar = reg_addr;
+	}
+
+	return tbar;
+}
+
+static u32 gsw_mdio_read(struct gswip_priv *priv, void *addr)
+{
+	struct mdio_device *mdio;
+	u32 reg_addr, tbar, val;
+
+	mdio = ((struct gsw_mdio *)dev_get_drvdata(priv->dev))->mdio_dev;
```

Comment:
The amount of dereferencing necessary to get the `struct mdio_device*` is the downside of this approach to structuring the driver data. It works, but I would want something more direct if we were in a performance-critical situation.

Reply:
A macro might be acceptable to the kernel devs if you're doing it a bunch. I wouldn't worry too much about performance if other drivers also do this.